feat: 뒹글 피드 조회 기능 추가

rugii913 opened the pull request:
## 연관 이슈
- closes #22 

## 해당 작업을 한 동기/이유는 무엇인가요?
- 우리 서비스의 기본적인 기능인 뒹글(익명 질문)들을 커서 기반 페이징을 통해 조회할 수 있는 기능을 추가했습니다.

## 리뷰어에게 작업 또는 변경 내용을 상세히 설명해주세요
- [x] 뒹글들(뒹글 피드)을 조회하는 데에 필요한 API 흐름 작성
    - web layer, service layer, persistence layer

- 참고사항 및 후속 작업
  - *application-local.yml 이름을 application-local-memory-h2로 변경했습니다.*
    - 이와 더불어 application-local-h2.yml 파일을 추가했습니다.
    - 디스크에 저장하는 H2 DB를 사용하려면 local-h2 프로파일로 실행하고, H2 인메모리 DB를 사용하려면 local-memory-h2 프로파일로 실행하면 됩니다.
  - 현재 DooingleController와 별개로 DooingleFeedController를 분리해둔 상태입니다. (API URL 문제)
    - API 명세에 대한 정리 작업 후 컨트롤러를 합치는 것을 생각해보겠습니다.
  - 스프링 데이터 제공 Pageable, Slice 대신 커서 기반 페이징에 특화된 클래스를 만들자는 의견이 있었습니다.
    - 다음 주에 리팩토링 작업 진행하겠습니다.

rugii913 committed:
feat: 뒹글 피드 조회, 팔로우하는 뒹글러 뒹글 피드 조회 API 흐름 추가

- 뒹글 피드 조회, 팔로우하는 뒹글러 뒹글 피드 조회 API 흐름 추가
  - 로직은 작성하지 않았음

rugii913 committed:
feat: PageRequest 작업 시작, 팔로우 관련 주석 처리, QClass 추가

- PageRequest 작업 시작
- 팔로우 관련 주석 처리
- QClass 추가
- (cf.) 컴파일 오류 때문에 DooingleCountQueryDslRepositoryImpl 쪽 코드 변경했음, 작업자와 상의할 것
  - PR 시 충돌 유의

rugii913 committed:
feat: 뒹글 피드 조회 QueryDSL 메서드 초안 작성

- 뒹글 피드 조회 QueryDSL 메서드 초안 작성

rugii913 committed:
refactor: 뒹글 피드 내림차순 정렬 및 service에서 값 반환하도록 함

- 뒹글 피드 내림차순 정렬 및 service에서 값 반환하도록 함

rugii913 committed:
refactor: API URL 정정, cursor에 null 허용

- API URL 정정: 명세와 다르게 /api가 빠져있던 부분 추가
- cursor에 null 허용: null이 들어올 경우 가장 최근 글부터 보여줌

rugii913 committed:
refactor: join, fetch join을 이용하여 쿼리 수 줄이기

- join, fetch join을 사용하여 쿼리 수를 줄임
- (cf.) alias 주는 방법 참고 https://hyeon9mak.github.io/query-dsl-same-table-join/

rugii913 committed:
refactor: projection을 사용하여 불필요한 join을 줄임

- projection을 사용하여 불필요한 join을 줄임

rugii913 committed:
chore: 스프링 실행 구성

- 스프링 실행 구성
  - local-h2 프로파일: H2 디스크 사용 DB
  - local-h2-memory: H2 인메모리 DB

rugii913 committed:
Merge remote-tracking branch 'origin/dev' into feat/dooingle-feeds

rugii913 committed:
chore: 스프링 실행 구성 구성

- application-local-h2에 prefix = "dooingler.list.size"와 관련된 값들 추가
- application-local-memory-h2에 eof 추가

rugii913 committed:
chore: 스프링 실행 구성 수정-H2 DB MySQL 호환 모드 설정

- H2 DB MySQL 호환 모드 설정

adorecamus commented on src/main/kotlin/com/dooingle/domain/dooingle/repository/DooingleQueryDslRepositoryImpl.kt:
```diff
@@ -0,0 +1,42 @@
+    private val dooingle = QDooingle.dooingle
+    private val owner = QUser("ow")
+
+    override fun getDooinglePageable(cursor: Long?, pageable: Pageable): Slice<DooingleResponse> {
```
메서드 이름에 의문이 있습니다..!! 🥹 지금 이름은 뒹글pageable을 가져온다는 뜻처럼 보이는데..! getDooinglesByPageable 같은 이름은 어떠신가요..??!

rugii913 replied:
"뒹글pageable을 가져온다는 뜻" 자체를 의도한 게 맞고, 그 쪽이 더 낫다고 생각해서 붙인 이름입니다.
뒹글pageable 자체가 데이터를 불러오는 한 단위라고 생각했었는데, 직관적으로 와닿지 않는 이름인가요?

adorecamus replied:
음.. 그런 의미라면 getDooingleSlice여야 하지 않을까요??
Pageable은 페이지네이션에 필요한 값을 넣은 인터페이스이고(구현체 PageRequest)
페이지네이션 결과를 반환하는 애들은 Page(PageImpl), Slice(SliceImpl)이니까..??

rugii913 replied:
반환 타입 부분에서는 Slice로 되어있기 때문에 그렇게 생각할 수도 있을 것 같긴 합니다.
~다만 Slice도 Pageable의 subtype인 것을 생각하긴 했었는데요...~(잘못 알고 있었던 부분이네요)
아무튼 프로젝트에서 이용할 커스텀한 페이지 DTO를 바로 만들까 생각이 드네요.

adorecamus replied:
저는 찬성합니다! Slice에 Pageable 도 들어가있고 기타 불필요한(no-offset 기반 페이지네이션에 불필요한..) 값들이 포함되어 있어서 결과 list랑 hasNext 만 있어도 괜찮지 않을까 싶습니당..

rugii913 replied:
따로 이슈 파서 고민해보겠습니당

adorecamus commented on src/main/kotlin/com/dooingle/domain/dooingle/repository/DooingleQueryDslRepositoryImpl.kt:
```diff
@@ -0,0 +1,42 @@
+                )
+            )
+            .from(dooingle)
+            .join(dooingle.owner, owner) // fetchJoin() 사용하면 에러 발생
```
entity를 조회하는 게 아니라 dto로 바로 반환해서 fetchJoin을 사용 못 하는 게 아닐지... 궁예해봅니다..! 🥹

rugii913 replied:
fetchJoin()이 있는 경우 발생하는 에러 메시지
2024-03-04T12:05:11.570+09:00 ERROR 9408 --- [nio-8080-exec-9] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: org.springframework.dao.InvalidDataAccessApiUsageException: org.hibernate.query.SemanticException: Query specified join fetching, but the owner of the fetched association was not present in the select list [SqmSingularJoin(com.dooingle.domain.dooingle.model.Dooingle(dooingle).owner(ow) : owner)]] with root cause

org.hibernate.query.SemanticException: Query specified join fetching, but the owner of the fetched association was not present in the select list [SqmSingularJoin(com.dooingle.domain.dooingle.model.Dooingle(dooingle).owner(ow) : owner)]
	at org.hibernate.query.sqm.tree.select.SqmQuerySpec.assertFetchOwner(SqmQuerySpec.java:573) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.query.sqm.tree.select.SqmQuerySpec.validateFetchOwners(SqmQuerySpec.java:548) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.query.sqm.tree.select.SqmQuerySpec.validateFetchOwners(SqmQuerySpec.java:494) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.query.sqm.tree.select.SqmQuerySpec.validateQueryStructureAndFetchOwners(SqmQuerySpec.java:469) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitStatement(SemanticQueryBuilder.java:403) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.query.hql.internal.SemanticQueryBuilder.buildSemanticModel(SemanticQueryBuilder.java:311) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.query.hql.internal.StandardHqlTranslator.translate(StandardHqlTranslator.java:71) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.query.internal.QueryInterpretationCacheStandardImpl.createHqlInterpretation(QueryInterpretationCacheStandardImpl.java:165) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.query.internal.QueryInterpretationCacheStandardImpl.resolveHqlInterpretation(QueryInterpretationCacheStandardImpl.java:147) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.internal.AbstractSharedSessionContract.interpretHql(AbstractSharedSessionContract.java:790) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:840) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:753) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:136) ~[hibernate-core-6.4.4.Final.jar:6.4.4.Final]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.orm.jpa.ExtendedEntityManagerCreator$ExtendedEntityManagerInvocationHandler.invoke(ExtendedEntityManagerCreator.java:364) ~[spring-orm-6.1.4.jar:6.1.4]
	at jdk.proxy2/jdk.proxy2.$Proxy132.createQuery(Unknown Source) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.orm.jpa.SharedEntityManagerCreator$SharedEntityManagerInvocationHandler.invoke(SharedEntityManagerCreator.java:319) ~[spring-orm-6.1.4.jar:6.1.4]
	at jdk.proxy2/jdk.proxy2.$Proxy132.createQuery(Unknown Source) ~[na:na]
	at com.querydsl.jpa.impl.AbstractJPAQuery.createQuery(AbstractJPAQuery.java:132) ~[querydsl-jpa-5.0.0-jakarta.jar:na]
	at com.querydsl.jpa.impl.AbstractJPAQuery.createQuery(AbstractJPAQuery.java:125) ~[querydsl-jpa-5.0.0-jakarta.jar:na]
	at com.querydsl.jpa.impl.AbstractJPAQuery.fetch(AbstractJPAQuery.java:242) ~[querydsl-jpa-5.0.0-jakarta.jar:na]
	at com.dooingle.domain.dooingle.repository.DooingleQueryDslRepositoryImpl.getDooinglesBySlice(DooingleQueryDslRepositoryImpl.kt:38) ~[main/:na]

rugii913 replied:
이 부분은 프로젝트 진행과는 관련 없는 부분이어서 지금 깊게 파보진 않겠습니다.

adorecamus replied:
https://velog.io/@yoho98/QueryDSL-%EC%97%90%EB%9F%AC-query-specified-join-fetching-but-the-owner-of-the-fetched-association-was-not-present-in-the-select-list

dto를 반환하도록 조회해서 그런 게 맞는 것 같습니당 핳ㅎ..

adorecamus commented on src/main/kotlin/com/dooingle/domain/dooingle/repository/DooingleQueryDslRepositoryImpl.kt:
```diff
@@ -0,0 +1,42 @@
+            .join(dooingle.owner, owner) // fetchJoin() 사용하면 에러 발생
+            .where(lessThanCursor(cursor))
+            .orderBy(dooingle.id.desc())
+            .limit(10) // TODO 10을 변수에 저장할 방법 생각하기
```
저는 ConfigurationProperties 클래스를 만들고 yml 파일에 써주는 식으로 했는데.. 그건 어떠신가요?!!
근데 피드인데 10 너무 적지 않습니까?!!

rugii913 replied:
한 번에 가져오는 수는 20으로 늘리겠습니다.
그런데 이 값을 스프링 실행 환경 변수로 관리하는 것은 뭔가 좀... 제가 근거를 정확하게 들기는 어려운데 안 좋은 직감이 듭니다.
혹시 참고한 자료 같은 것이 있으신가요??

adorecamus replied:
아뇨 아무 생각 없이 썼습니다...😭
저도 이슈 올려서 수정해야겠네요...
그러면 enum 클래스나 object를 만들어서 거기서 다 같이 저장하는 건 어떠세요??
DOOINGLE_FEED_SIZE
HOT_DOOINGLER_LIST_SIZE
NEW_DOOINGLER_LIST_SIZE
이런 식으루..?

rugii913 replied:
일단 20이라는 숫자는 pageale에서 가져왔습니다. 간단한 방법을 못 떠올리고 있었네요.
그리고 controller에서 companion object로 const val PAGE_SIZE = 20을 갖고 있습니다.
이 부분이 적절할지는 더 고민해볼 필요가 있을 것 같습니다.

adorecamus commented on src/main/kotlin/com/dooingle/domain/dooingle/repository/DooingleQueryDslRepositoryImpl.kt:
```diff
@@ -0,0 +1,42 @@
+            .orderBy(dooingle.id.desc())
+            .limit(10) // TODO 10을 변수에 저장할 방법 생각하기
+            .fetch()
+            .let { SliceImpl(it) }
```
생성자에 content만 담으면 무조건 hasNext 가 false라서..
public SliceImpl(List<T> content) {
        this(content, Pageable.unpaged(), false);
    }

요 생성자를 쓰셔야 할 것 같습니다!!
public SliceImpl(List<T> content, Pageable pageable, boolean hasNext) {
        super(content, pageable);
        this.hasNext = hasNext;
        this.pageable = pageable;
    }

그리고 hasNext 를 판단하는 로직도 필요할 것 같습니당 🥹

rugii913 replied:
hasNext 로직 추가했습니다~

adorecamus reviewed (approved):
고생하셨습니다!! 필요한 로직이 있는 것 같아서 한 번 확인해보시구 반영해서 올려주세용..!! 🥹

rugii913 committed:
refactor: 메서드 이름 수정, 페이지 사이즈 조정, hasNext 판단 로직 추가

- 메서드 이름 수정
  - 가져오는 대상이 Pageable이 아니라 Slice이므로 이름이 잘못된 부분을 수정
- 페이지 사이즈 조정
  - 페이지 사이즈가 작다는 의견이 있어 조정
- hasNext 판단 로직 추가
  - 프론트 작업을 위해 Slice에 hasNext 정보 추가

rugii913 committed:
chore: 프로파일 없어서 테스트 깨지는 부분 수정

- 테스트 깨지는 부분 수정
  - 테스트에서 SpringBoot가 제대로 실행되지 않아 테스트가 실패하고 있었음
  - SpringBoot를 시작하기 위한 환경 구성 파일 추가, @ ActiveProfiles 추가

rugii913 committed:
add: 뒹글 피드 슬라이스의 내용물 20개만 반환

- 뒹글 피드 슬라이스의 내용물 20개만 반환
  - hasNext를 판단하기 위해 쿼리할 때에는 21개 뒹글을 요청하고 있었음
  - 21개 뒹글을 받아온 뒤 마지막 뒹글을 없애야 원하는 대로 20개의 뒹글만 응답으로 보낼 수 있음

rugii913 commented:
- 뒹글 피드 슬라이스 내용물을 21개 반환하고 있던 부분을 수정하여, 응답으로는 20개만 반환하도록 했습니다.
- 커서 기반 페이징에 특화된 클래스 작성 이슈 등록했습니다.

adorecamus commented on src/main/kotlin/com/dooingle/domain/dooingle/repository/DooingleQueryDslRepositoryImpl.kt:
```diff
@@ -30,12 +30,12 @@ class DooingleQueryDslRepositoryImpl(
             .limit(selectSize.toLong())
             .fetch()
-            .let { SliceImpl(it, pageable, hasNextSlice(it, selectSize)) }
+            .let { SliceImpl(it.dropLast(1), pageable, hasNextSlice(it, selectSize)) }
```
오 dropLast 메서드 알아갑니다 🥹👍
